### PR TITLE
[JENKINS-59540] Add annotation to getUrl

### PR DESCRIPTION
**Pull request by res0nance**
## [JENKINS-59540](https://issues.jenkins-ci.org/browse/JENKINS-59540) Null check getUrl references

Make it clear to callers that getUrl may return null.

Adding an annotation for what I suspect was the cause of [JENKINS-50014](https://issues.jenkins-ci.org/browse/JENKINS-50014). Filed jenkinsci/git-parameter-plugin#83 against that plugin but circling back here to ensure that other plugins can catch this issue with the annotation added.

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.md) doc
- [x] I have referenced the Jira issue related to my changes in one or more commit messages
- [ ] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No spotbugs warnings were introduced with my changes

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: src/main/java/jenkins/plugins/git/GitSCMFileSystem.java [hunk 1]
```diff
@@ -282,13 +282,16 @@ public SCMFileSystem build(@NonNull Item owner, @NonNull SCM scm, @CheckForNull
             if (rev != null && !(rev instanceof AbstractGitSCMSource.SCMRevisionImpl)) {
                 return null;
             }
-            TaskListener listener = new LogTaskListener(LOGGER, Level.FINE);
             GitSCM gitSCM = (GitSCM) scm;
             UserRemoteConfig config = gitSCM.getUserRemoteConfigs().get(0);
             BranchSpec branchSpec = gitSCM.getBranches().get(0);
             String remote = config.getUrl();
+            if (remote == null) {
```
Comment (MarkEWaite) [inline src/main/java/jenkins/plugins/git/GitSCMFileSystem.java]: Please log a warning message that the URL in a null URL was configured int he UserRemoteConfig.  Silently returning null seems like it will need diagnosis by the administrator or the user.

**Comment by res0nance** [comment]
Spotbugs has been fixed. Its nice that one discovery leads to another and further leads to several bugs being fixed.

File: src/main/java/jenkins/plugins/git/GitSCMFileSystem.java (hunk 1 above)
Comment (res0nance) [reply src/main/java/jenkins/plugins/git/GitSCMFileSystem.java]: Added logging

**Comment by res0nance** [comment]
@MarkEWaite is there anything else needed here?

File: src/main/java/jenkins/plugins/git/GitSCMFileSystem.java [hunk 2]
```diff
@@ -282,11 +282,16 @@ public SCMFileSystem build(@NonNull Item owner, @NonNull SCM scm, @CheckForNull
             if (rev != null && !(rev instanceof AbstractGitSCMSource.SCMRevisionImpl)) {
                 return null;
             }
-            TaskListener listener = new LogTaskListener(LOGGER, Level.FINE);
             GitSCM gitSCM = (GitSCM) scm;
             UserRemoteConfig config = gitSCM.getUserRemoteConfigs().get(0);
             BranchSpec branchSpec = gitSCM.getBranches().get(0);
             String remote = config.getUrl();
+            LogTaskListener listener = new LogTaskListener(LOGGER, Level.FINE);
```
Comment (MarkEWaite) [inline src/main/java/jenkins/plugins/git/GitSCMFileSystem.java]: Why the change from TaskListener to LogTaskListener as the type?

File: src/main/java/jenkins/plugins/git/GitSCMFileSystem.java [hunk 3]
```diff
@@ -282,11 +282,16 @@ public SCMFileSystem build(@NonNull Item owner, @NonNull SCM scm, @CheckForNull
...
+            if (remote == null) {
+                listener.getLogger().println("Git remote url is null");
+                listener.close();
```
Comment (MarkEWaite) [inline src/main/java/jenkins/plugins/git/GitSCMFileSystem.java]: If close is needed here before the return from the method, shouldn't it also be done at every other return from the method?

Comment (res0nance) [reply src/main/java/jenkins/plugins/git/GitSCMFileSystem.java]: I think i did it because my IDE was warning on that. I'll remove that

File: src/main/java/jenkins/plugins/git/GitSCMFileSystem.java (hunk 2 above)
Comment (res0nance) [reply src/main/java/jenkins/plugins/git/GitSCMFileSystem.java]: Didn't notice the type changed I'll revert that bit
